### PR TITLE
Add feature to make applicable local functions static.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Looks like this:

image

Todo:
- [x] Tests
- [x] User facing option in Tools|Options

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @mavasani

File: src/EditorFeatures/CSharpTest/MakeLocalFunctionStatic/MakeLocalFunctionStaticTests.cs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        static async Task<int> fibonacci(int n)
+        {
+            return n <= 1 ? n : await fibonacci(n - 1) + await fibonacci(n - 2);
+        }
+    }
+}",
+parseOptions: CSharp8ParseOptions);
+        }
+    }
+}
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/MakeLocalFunctionStatic/MakeLocalFunctionStaticTests.cs]: Consider testing nested local function

File: src/Features/Core/Portable/FeaturesResources.Designer.cs [hunk 2]
```diff
@@ -19,7 +19,7 @@ namespace Microsoft.CodeAnalysis {
     // class via a tool like ResGen or Visual Studio.
     // To add or remove a member, edit your .ResX file then rerun ResGen
     // with the /str option, or rebuild your VS project.
-    [global::System.CodeDom.Compiler.GeneratedCodeAttribute("System.Resources.Tools.StronglyTypedResourceBuilder", "15.0.0.0")]
```
Comment (jcouv) [inline src/Features/Core/Portable/FeaturesResources.Designer.cs]: >"15. [](start = 115, length = 4)

Please restore so we don't flip/flop

File: src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionCodeFixProvider.cs [hunk 3]
```diff
@@ -75,20 +75,26 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
             var root = editor.OriginalRoot;
             var currentRoot = root.TrackNodes(nodesToTrack);
 
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var languageVersion = ((CSharpParseOptions)semanticModel.SyntaxTree.Options).LanguageVersion;
+            var makeStaticIfPossible = languageVersion >= LanguageVersion.CSharp8 &&
+                options.GetOption(CSharpCodeStyleOptions.PreferStaticLocalFunction).Value;
```
Comment (jcouv) [inline src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionCodeFixProvider.cs]: >PreferStaticLocalFunction [](start = 57, length = 25)

I probably wouldn't have added an option for this. `static` is better :-P

File: src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionCodeFixProvider.cs [hunk 4]
```diff
@@ -100,12 +106,41 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
...
         }
 
+        private static bool MakeStatic(
+            SemanticModel semanticModel,
+            bool makeStaticIfPossible,
+            LocalDeclarationStatementSyntax localDeclaration,
+            CancellationToken cancellationToken)
+        {
+            // Determines if we can make the local function 'static'.  We can make it static
+            // if the original lambda did not cpature any variables (other than the local
```
Comment (jcouv) [inline src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionCodeFixProvider.cs]: >cpature [](start = 46, length = 7)

typo

File: src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionCodeFixProvider.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionCodeFixProvider.cs]: i'm worried about any option people can't turn off.  If this starts just telling them to change, and they prefer the briefer form, i don't want them to be unhappy :)

File: src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionCodeFixProvider.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionCodeFixProvider.cs]: intentional.  it verified you're paying attention

**Comment by jcouv** [comment]
Thanks @CyrusNajmabadi!  @cston FYI :-)

File: src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionCodeFixProvider.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionCodeFixProvider.cs]: Fixed in the useSimpleUsingDeclaration branch.

**Comment by CyrusNajmabadi** [comment]
@genlu @jasonmalinowski  As usual, i hate that `master` isn't targeting the latest release. :)

Do you know when it will?  If it's going to be targeting the latest release in a few days, i'll just wait.  If it's going to be a while, i'll rebase.  Thanks!

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi You should be able to just retarget this into dev16.1-preview1 without any hard work -- just edit your PR target.

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski Ok.  Retargetted.  Feel free to merge when you can.

File: src/Features/CSharp/Portable/MakeLocalFunctionStatic/MakeLocalFunctionStaticCodeRefactoringProvider.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.CSharp.MakeLocalFunctionStatic
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = nameof(MakeLocalFunctionStaticCodeFixProvider)), Shared]
+    internal class MakeLocalFunctionStaticCodeFixProvider : SyntaxEditorBasedCodeFixProvider
```
Comment (genlu) [inline src/Features/CSharp/Portable/MakeLocalFunctionStatic/MakeLocalFunctionStaticCodeRefactoringProvider.cs]: type name and file name not in sync.

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/MakeLocalFunctionStatic/MakeLocalFunctionStaticCodeRefactoringProvider.cs]: Fixed.  thanks!

**Comment by CyrusNajmabadi** [comment]
@genlu @jcouv @jinujoseph this is ready to go into dev16.1-preview1

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**